mm2src/coins/eth.rs

```diff
@@ -2431,6 +2437,7 @@ async fn sign_transaction_with_keypair(
     Ok((
         tx.sign(key_pair.secret(), coin.chain_id),
         web3_instances_with_latest_nonce,
+        nonce_lock,
     ))
 }
```

Comment:
what difference does it make if we just drop the `nonce_lock` before returning from this function and returning it without using e.g

```rust 
async fn sign_transaction_with_keypair(
    ctx: MmArc,
    coin: &EthCoin,
    key_pair: &KeyPair,
    value: U256,
    action: Action,
    data: Vec<u8>,
    gas: U256,
) -> Result<(SignedEthTx, Vec<Web3Instance>), TransactionErr> {
    let mut status = ctx.log.status_handle();
    macro_rules! tags {
        () => {
            &[&"sign"]
        };
    }
    let nonce_lock = coin.nonce_lock.lock().await;
    status.status(tags!(), "get_addr_nonce…");
    let (nonce, web3_instances_with_latest_nonce) =
        try_tx_s!(coin.clone().get_addr_nonce(coin.my_address).compat().await);
    status.status(tags!(), "get_gas_price…");
    let gas_price = try_tx_s!(coin.get_gas_price().compat().await);

    let tx = UnSignedEthTx {
        nonce,
        gas_price,
        gas,
        action,
        value,
        data,
    };

    let _ = nonce_lock;
    Ok((
        tx.sign(key_pair.secret(), coin.chain_id),
        web3_instances_with_latest_nonce,
    ))
}
```

Reply:
A nonce can not be reused in another tx. We get the nonce from RPC nodes (we pick the highest of them) and use it in out tx.
If another tx is being created at the same time from another thread (which is the case tested by `test_nonce_lock` test) they would get the same nonce from the RPC. And the eth network will reject the second one.

We will want to return the lock so the caller can have the nonce locked until the transaction is relayed successfully and the nonce is updated on the connected RPC nodes.

Reply:
ok thanks for the explanation 🙏🏼